repo/fsrepo/config_test.go

```diff
@@ -234,3 +246,81 @@ func TestMeasureConfig(t *testing.T) {
 		t.Errorf("expected '*measure.measure' got '%s'", typ)
 	}
 }
+
+func TestBadgerConfig(t *testing.T) {
+	config := new(config.Datastore)
+	err := json.Unmarshal(defaultConfig, config)
+	if err != nil {
+		t.Fatal(err)
+	}
+	dir, err := ioutil.TempDir("", "ipfs-datastore-config-test")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer os.RemoveAll(dir) // clean up
+
+	spec := make(map[string]interface{})
+	err = json.Unmarshal(badgerConfig, &spec)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	dsc, err := fsrepo.AnyDatastoreConfig(spec)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	expected := `{"path":"datastore","type":"badgerds"}`
+	if dsc.DiskSpec().String() != expected {
+		t.Errorf("expected '%s' got '%s' as DiskId", expected, dsc.DiskSpec().String())
+	}
+
+	ds, err := dsc.Create(dir)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if typ := reflect.TypeOf(ds).String(); typ != "*badger.Datastore" {
+		t.Errorf("expected '*badger.datastore' got '%s'", typ)
+	}
+}
+
+func TestBadger2Config(t *testing.T) {
+	config := new(config.Datastore)
+	err := json.Unmarshal(defaultConfig, config)
+	if err != nil {
+		t.Fatal(err)
+	}
+	dir, err := ioutil.TempDir("", "ipfs-datastore-config-test")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer os.RemoveAll(dir) // clean up
+
+	spec := make(map[string]interface{})
+	err = json.Unmarshal(badger2Config, &spec)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	dsc, err := fsrepo.AnyDatastoreConfig(spec)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	expected := `{"path":"datastore","type":"badger2ds"}`
+	if dsc.DiskSpec().String() != expected {
+		t.Errorf("expected '%s' got '%s' as DiskId", expected, dsc.DiskSpec().String())
+	}
+
+	ds, err := dsc.Create(dir)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// The name of the package in go-ds-badger2 is "badger", so the type
+	// returned by dsc.Create() is "*badger.Datastore"
+	if typ := reflect.TypeOf(ds).String(); typ != "*badger.Datastore" {
```

Comment:
would using reflect.Typeof(ds).PkgPath() help disambiguate here?

Reply:
Unfortunately `PkgPath()` always returns `""` here.  I cannot yet find a clear explanation why, but it appears to do with the way interfaces are embedded.  However, if we phase-out badger1 then this will no longer be an issue.